File: src/python/portfolio_management/online/db2_recovery.py
"""DB2 Recovery Manager for Online Programs - migrated from DB2RECV.cbl.

Handles DB2 connection failures, implements retry logic, manages
transaction rollback, and provides recovery status tracking.
"""

import logging
import time
from dataclasses import dataclass
from datetime import datetime
from typing import Optional

from portfolio_management.models.common import ReturnCode

logger = logging.getLogger(__name__)

PROGRAM_ID = "DB2RECV"
MAX_RETRIES = 3
RETRY_INTERVAL_SECONDS = 2


@dataclass
class RecoveryStatus:
    recovery_type: str = ""
    attempts: int = 0
    success: bool = False
    last_error: str = ""
    timestamp: str = ""


class DB2RecoveryManager:
    def __init__(self):
        self._recovery_history: list[RecoveryStatus] = []

    def recover_connection(self, connection_factory: Optional[type] = None) -> int:
        status = RecoveryStatus(
            recovery_type="CONNECTION",
            timestamp=datetime.now().strftime("%Y-%m-%d-%H.%M.%S.%f"),
        )

        for attempt in range(1, MAX_RETRIES + 1):
            status.attempts = attempt
            try:
                if connection_factory is not None:
                    connection_factory()
                status.success = True
                logger.info("Connection recovered on attempt %d", attempt)
                self._recovery_history.append(status)
                return ReturnCode.SUCCESS
            except Exception as e:
                status.last_error = str(e)
                logger.warning("Connection recovery attempt %d failed: %s", attempt, e)
                if attempt < MAX_RETRIES:
                    time.sleep(RETRY_INTERVAL_SECONDS)

        logger.error("Connection recovery failed after %d attempts", MAX_RETRIES)
        self._recovery_history.append(status)
        return ReturnCode.ERROR

    def recover_transaction(self, rollback_func: Optional[type] = None) -> int:
        status = RecoveryStatus(
            recovery_type="TRANSACTION",
            timestamp=datetime.now().strftime("%Y-%m-%d-%H.%M.%S.%f"),
            attempts=1,
        )

        try:
            if rollback_func is not None:
                rollback_func()
            status.success = True
            logger.info("Transaction rolled back successfully")
            self._recovery_history.append(status)
            return ReturnCode.SUCCESS
        except Exception as e:
            status.last_error = str(e)
            logger.error("Transaction recovery failed: %s", e)
            self._recovery_history.append(status)
            return ReturnCode.ERROR

    def recover_cursor(self, cursor_manager: Optional[object] = None, cursor_name: str = "") -> int:
        status = RecoveryStatus(
            recovery_type="CURSOR",
            timestamp=datetime.now().strftime("%Y-%m-%d-%H.%M.%S.%f"),
            attempts=1,
        )

        try:
            if cursor_manager is not None and hasattr(cursor_manager, "close"):
                cursor_manager.close(cursor_name)
            status.success = True
            logger.info("Cursor %s recovered", cursor_name)
            self._recovery_history.append(status)
            return ReturnCode.SUCCESS
        except Exception as e:
            status.last_error = str(e)
            logger.error("Cursor recovery failed: %s", e)
            self._recovery_history.append(status)
            return ReturnCode.ERROR

    def get_recovery_history(self) -> list[RecoveryStatus]:
        return list(self._recovery_history)

File: ts-src/programs/online/db2-recovery.ts
/**
 * Online DB2 Recovery Handler.
 * Migrated from: src/programs/online/DB2RECV.cbl
 *
 * Handles connection recovery, transaction recovery, and cursor recovery
 * for the online layer.
 */

import { Knex } from 'knex';
import { connectToDatabase, checkDatabaseStatus } from '../../database';
import { ReturnCode } from '../../types';

export type RecoveryType = 'CONNECTION' | 'TRANSACTION' | 'CURSOR';

export class Db2Recovery {
  private readonly maxRetries = 3;
  private readonly retryDelayMs = 1000;

  constructor(private readonly db: Knex) {}

  /** Dispatch recovery action. */
  async recover(type: RecoveryType): Promise<number> {
    switch (type) {
      case 'CONNECTION':
        return this.recoverConnection();
      case 'TRANSACTION':
        return this.recoverTransaction();
      case 'CURSOR':
        return this.recoverCursor();
      default:
        return ReturnCode.Error;
    }
  }

  /** 1000-RECOVER-CONNECTION – re-establish database connection. */
  private async recoverConnection(): Promise<number> {
    for (let attempt = 1; attempt <= this.maxRetries; attempt++) {
      try {
        const isOk = await checkDatabaseStatus();
        if (isOk) {
          console.log('Connection recovery: connection is active');
          return ReturnCode.Success;
        }

        console.log(`Connection recovery attempt ${attempt}/${this.maxRetries}`);
        await connectToDatabase();
        return ReturnCode.Success;
      } catch (err) {
        console.error(`Connection recovery attempt ${attempt} failed: ${err}`);
        if (attempt < this.maxRetries) {
          await this.sleep(this.retryDelayMs * attempt);
        }
      }
    }

    console.error('Connection recovery failed after all retries');
    return ReturnCode.Severe;
  }

  /** 2000-RECOVER-TRANSACTION – rollback any pending transaction. */
  private async recoverTransaction(): Promise<number> {
    try {
      // Attempt to rollback any pending work
      await this.db.raw('ROLLBACK');
      console.log('Transaction recovery: rolled back pending work');
      return ReturnCode.Success;
    } catch {
      // No pending transaction to rollback – that's OK
      return ReturnCode.Success;
    }
  }

  /** 3000-RECOVER-CURSOR – cursors are in-memory, just log. */
  private recoverCursor(): number {
    console.log('Cursor recovery: in-memory cursors cleared on reconnect');
    return ReturnCode.Success;
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}
